_posts/2021-04-02-being-productive-with-typescript.md

```diff
@@ -0,0 +1,477 @@
+Frankly for everything not coming from an external source, I probably would use a class.
+Yet there are still exceptions to this, and every case should be handled pragmatically.
+But for this post I am working with interfaces.
+And without any further ado, let's dive into some code.
+
+# Literal narrowing
+```typescript
+interface AgendaItem {
+  status: string;
+}
+
+function pickColor(status: string) {
+  switch (status){
+    case 'OPEN':
+      return '#2266FF'
+    case 'CLOSED':
+      return '#888888'
+    case 'IN PROGRESS':
+      return '#FF8800'
+    }
+}
+```
+
+The code above seems okay. 
+I used to write like this all the time. 
+But the question is: **What is the added value of Typescript?** 
+Let us make a list with what Typescript does for us here:
+
+- ~~Readability~~
+- ~~Code hinting~~
+- ~~Compile-time errors~~
+
+Needless to say, this code isn't any better than its Javascript counterpart. 
+Let's have a look at an improved version.
+
+```typescript
+type AgendaItemStatus = 'OPEN' | 'CLOSED' | 'IN PROGRESS' | 'DELETED';
+
+interface AgendaItem {
+ status: AgendaItemStatus;
+}
+
+function pickColor(status: AgendaItemStatus) {
+ switch (status){
+   case 'OPEN':
+     return '#2266FF'
+   case 'CLOSED':
+     return '#888888'
+   case 'IN PROGRESS':
+     return '#FF8800'
+   case 'TODO':
+     return '#FF8800'
+ }
+}
+```
+Consider the new type: `AgendaItemStatus`. 
+It is called a [string literal](https://www.typescriptlang.org/docs/handbook/literal-types.html#string-literal-types){:target="_blank" rel="noopener noreferrer"} and should have all possible statuses listed.
+This makes my code more typesafe and gives the code hinting we would expect from Typescript.
+In the example above there is a case for `'TODO'`, but thanks to Typescript your IDE knows `'TODO'` is not a possible status.
+It gets even better, some IDEs allow you to autogenerate the switch case.
+
+Personally I use a [string literal](https://www.typescriptlang.org/docs/handbook/literal-types.html#string-literal-types){:target="_blank" rel="noopener noreferrer"} when I can easily remember the statuses without looking it up in a database. 
+On average I can remember around 7 different statuses.
+
+Let us have a look at our list of added Typescript values retrospectively:
+- Readability
+- Code hinting
+- ~~Compile-time errors~~
+
+How are we missing out on the compile-time errors?
+Imagine my data source adds a new status. 
+I will add the new status to the list, but no mechanism is forcing me to alter the `pickColor` function.
+Code hinting will not always help me out here, so I need a compile-time error to catch this as soon as possible.
+
+```typescript
+type AgendaItemStatus = 'OPEN' | 'CLOSED' | 'IN PROGRESS' | 'DELETED';
+
+interface AgendaItem {
+ status: AgendaItemStatus;
+}
+
+type AgendaItemStatusMap<T> = Record<AgendaItemStatus, T>;
+// type AgendaItemStatusToColorMap<T> = { [K in AgendaItemStatus]: T }
```

Comment:
What is this comment doing here?

Reply:
It's another way (more expressive) to achieve the same, but I'am gonna drop it because it's only confusing probably